### ISSUE TITLE
MPI_Add_error_class limited number of calls?

### ISSUE THREAD

**Issue by dschwoerer** [pmodels/mpich#4341]
I am trying to call `PetscInitialize` several times (in unit tests) but it fails on the 128th time with:
```
Fatal error in MPI_Add_error_class: Other MPI error, error stack:
MPI_Add_error_class(95): MPI_Add_error_class(errorclass=0x7ffff3fad958) failed
MPI_Add_error_class(74): No more user-defined error classes
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=671687183
:
system msg for write_line failure : Bad file descriptor
[Inferior 1 (process 31862) exited with code 017]
```
openmpi has no issue with this.

The affected version is: `mpich-devel-3.2.1-9.fc30.x86_64`

**Comment by gcongiu** [comment]
@dschwoerer thank you for reporting this. It seems we statically allocate only 128 error classes in MPICH. As a quick fix can you try increasing the `MPIR_ERR_CLASS_SIZE` (and `MPIR_ERR_CLASS_MASK`) macro in `src/include/mpir_err.h` to the number of classes you need? Eventually, this should probably become a configurable parameter.

**Comment by pavanbalaji** [comment]
I think it's important to understand what the problem is before suggesting changing the number of error classes.  What is `PETScInitialize` doing to cause this?  Why are more classes being created?  Is it trying to reinitialize MPI without checking if it's already initialized?

@dschwoerer Do you have a simple MPI test (with no dependency on PETSc) that reproduces this?

**Comment by dschwoerer** [comment]
Thanks for the fast replies. PETSc is finalized, before it is initialised again. Thus it seems the PETSc behaviour is sound. I am not sure whether it is possible to remove error classes after they have been added, if so that could be another solution, but I don't know MPI that well.

I am perfectly happy with that number being finite, but does it need to be constant and a compile-time constant on top? If it would be dynamically allocated, the size could probably be several million to billion of calls before OOM kicks in.

PETSc is calling `MPI_Add_error_class` each time it is initialised. I haven't tried to implement a POC, but from this thread it seems calling `MPI_Add_error_class` for 129 times should do the job. It would probably be nice if that would be dynamically done, but I am happy for this to be closed as won't fix, and prioritize other issues.

**Comment by hzhou** [comment]
I don't think one is allowed to initialize MPI again once it has been finalized, which I believe PETSc does when it is finalized. If the unit-tests are designed basing on multiple MPI-Init/Finalize cycles, then the tests need be redesigned.

**Comment by ZedThree** [comment]
We handle MPI ourselves by calling `MPI_Init` before `PetscInitialize`, which causes PETSc to not finalise MPI in `PetscFinalize`. `MPI_Init/Finalize` are called only once during the whole program. We can hoist the PETSc init/finialise calls up to the testing `main()`, so this isn't an insurmountable problem.

Is there some way of resetting the state of the error classes? I couldn't see any way to remove/destroy error classes, codes, or strings in the standard.

**Comment by pavanbalaji** [comment]
@ZedThree You are right; the MPI standard does not give a way to remove error classes, which should probably be added.  Having said that, I still don't understand why you have to keep adding new error classes every time you initialize PETSc.  Are these completely new different error classes each time?

**Comment by hzhou** [comment]
> We handle MPI ourselves by calling `MPI_Init` before `PetscInitialize`, which causes PETSc to not finalise MPI in `PetscFinalize`.

I see. That makes sense. So it seems `PetscInitialize` is registering error classes each time. @pavanbalaji, looks like it would be useful to have an `MPI_Remove_error_class(int errorclass)` so PETSc can call that during `PetscFinalize`. @ZedThree, currently seems the solution is either not to Init/Finalize Petsc multiple times or patch Petsc not to register new error classes.

**Comment by hzhou** [comment]
> We handle MPI ourselves by calling `MPI_Init` before `PetscInitialize`, which causes PETSc to not finalise MPI in `PetscFinalize`.

I see. That makes sense. So it seems `PetscInitialize` is registering error classes each time. @pavanbalaji, looks like it would be useful to have an `MPI_Remove_error_class(int errorclass)` so PETSc can call that during `PetscFinalize`. @ZedThree, currently seems the solution is either not to Init/Finalize Petsc multiple times or patch Petsc not to register new error classes.

**Comment by ZedThree** [comment]
@pavanbalaji No, it's the same error class, it's just that PETSc can't tell it's already added it. It's handled by PETSc, we have no control over it.

@hzhou I've gone with the first option, but I wonder if the following would be suitable patch for PETSc:

```cpp
PetscMPIInt PETSC_MPI_ERROR_CLASS = MPI_ERR_LASTCODE;

...

if (PETSC_MPI_ERROR_CLASS == MPI_ERR_LASTCODE) {
  ierr = MPI_Add_error_class(&PETSC_MPI_ERROR_CLASS);CHKERRQ(ierr);
  ierr = MPI_Add_error_code(PETSC_MPI_ERROR_CLASS,&PETSC_MPI_ERROR_CODE);CHKERRQ(ierr);
}
```

If I understand correctly, `MPI_ERR_LASTCODE` is a constant and any user error class will be larger than this, so it's safe to use as a sentinel?

**Comment by pavanbalaji** [comment]
@ZedThree Adding a `MPI_Remove_error_class` is probably the right option here, but it'll take a long time to get that into MPI (we are too late for MPI-4).  The best option at this point is to have PETSc internally keep track of the fact that it already added the error class and not add it again.

**Comment by hzhou** [comment]
> @hzhou I've gone with the first option, but I wonder if the following would be suitable patch for PETSc:
> 
> ```c++
> PetscMPIInt PETSC_MPI_ERROR_CLASS = MPI_ERR_LASTCODE;
> 
> ...
> 
> if (PETSC_MPI_ERROR_CLASS == MPI_ERR_LASTCODE) {
>   ierr = MPI_Add_error_class(&PETSC_MPI_ERROR_CLASS);CHKERRQ(ierr);
>   ierr = MPI_Add_error_code(PETSC_MPI_ERROR_CLASS,&PETSC_MPI_ERROR_CODE);CHKERRQ(ierr);
> }
> ```
> 
> If I understand correctly, `MPI_ERR_LASTCODE` is a constant and any user error class will be larger than this, so it's safe to use as a sentinel?

`MPI_ERR_LASTCODE` is a constant, i.e. static and never change. Therefore, you can't use it as a sentinel.

The patch need be able to load a persistent state to inform `PetscInitialize` that the error class has already been registered and use that persistent error code instead.

**Comment by hzhou** [comment]
On second thought, the solution may not be so bad. Just let `Petsc_Initialize` accept an additional parameter or check an environment variable that sets to the error code it supposed to use. Then in your application, just pre-register the error class and pass the error code in to Petsc.

**Comment by ZedThree** [comment]
BTW, here's a reproducer:
```cpp
#include <mpi.h>
#include <petsc.h>

int main() {
  // Initialize the MPI environment
  MPI_Init(nullptr, nullptr);

  for (int i = 0; i < 128; ++i) {
    PetscInitialize(nullptr, nullptr, PETSC_NULL, nullptr);
    PetscFinalize();
  }

  // Finalize the MPI environment.
  MPI_Finalize();
}
```
Compiled with `mpic++ -std=c++11 petsc_loop.cxx -o petsc_loop $(pkg-config --cflags PETSc) $(pkg-config --libs PETSc)`. I've been using MPICH 3.2 just because it was the most convenient installed version I have access to.

I tried my patch above and it seemed to work, but there may be consequences I've not seen. I'll go talk to the PETSc devs, see what they think is acceptable.

**Comment by jedbrown** [comment]
The proposed patch looks fine to me. @hzhou What was your concern? The patch just depends on `MPI_ERR_LASTCODE` not changing. It could use an arbitrary predefined value (`MPI_ERR_UNKNOWN`, etc.) instead.

**Comment by hzhou** [comment]
> The proposed patch looks fine to me. @hzhou What was your concern? The patch just depends on `MPI_ERR_LASTCODE` not changing. It could use an arbitrary predefined value (`MPI_ERR_UNKNOWN`, etc.) instead.

I was thinking petsc library was reloaded in between -- a mis-thinking :)

**Comment by ZedThree** [comment]
I've opened a proposal for adding `MPI_Remove_error*` to the standard: https://github.com/mpi-forum/mpi-issues/issues/283 :smile:

**Comment by hzhou** [comment]
@ZedThree Is it OK to close this issue?

**Comment by ZedThree** [comment]
Yes, workaround implemented in our code, patch submitted to PETSc, and proposal submitted to MPI!
